### ISSUE TITLE
Implement decl and decl/binding for top-level fns [phab:D948]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1271]
_Created by **djrenren** at 2016-07-06 20:32:28_: 

This change introduces the ability to index function declarations. Bindings are tied to the name of the function while the declaration is the full block starting from a possible access modifier all the way through the name, and declaration body.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 18:26:22_:

No commented code, please.  Also, why not have the language set?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 18:26:22_:

Put language name into constant.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 18:26:22_:

Quick comment for each of these functions.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 18:26:22_:

More descriptive name?  I don't know what Pass is.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 18:26:22_:

`name = "all"`

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 18:26:22_:

These goals are not prefixed correctly: `%s#// -#//-#`

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-07-07 18:57:11_:

  - Comment Corpus, fix tests

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2016-07-07 19:08:28_:

Whoops, commented this out while debugging

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 20:51:57_:

```
//- AFile = vname("", "", "", "kythe/rust/indexer/testdata/basic/a.rs", "")
//-   .node/kind file
//- AFile.text "fn main(){}\n"
```

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 20:51:57_:

Be consistent with the string types used in this struct; pick one.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 20:51:57_:

Fact values do not need to UTF8.  This should probably be `&[u8]`.

Feel free to only add a `TODO(djrenren)` and handle this in a separate change.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 20:51:57_:

There's not a stderr version of `println!`?  Or some kind of logging library?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 20:51:57_:

Instead of using `unwrap`, this function should return a `Result<VName, ...>`.

**Comment by kythe** [comment]
_**schroederc** required changes at 2016-07-07 20:51:57_

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 22:22:12_:

Just `Ok(self.anchor(&start.fm.name, start_byte, end_byte))`

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 22:22:12_:

Please put in a TODO about logging.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-07 22:22:12_:

I would suggest going with `String` all the way.  It simplifies things and we don't yet have any concern about copying data.

**Comment by kythe** [comment]
_**djrenren** wrote at 2016-07-07 22:39:53_:

  - Make VName uniform Option<String> and add TODOs
  - Update TODO names to match github name

**Comment by kythe** [comment]
_**schroederc** accepted this revision at 2016-07-07 22:44:39_

**Comment by kythe** [comment]
_**djrenren** closed via commit fac7de65ab8cab80af9ceaacd616f11658df438f at 2016-07-07 23:03:17_
